explorer/scan/discovery.go

```diff
@@ -0,0 +1,191 @@
+	im, err := manager.NewManager(manager.WithInventory(inv, providers.DefaultRuntime()))
+	if err != nil {
+		return nil, errors.New("failed to resolve inventory for connection")
+	}
+	invAssets := im.GetAssets()
+	if len(invAssets) == 0 {
+		return nil, errors.New("could not find an asset that we can connect to")
+	}
+
+	runtimeEnv := execruntime.Detect()
+	var runtimeLabels map[string]string
+	// If the runtime is an automated environment and the root asset is CI/CD, then we are doing a
+	// CI/CD scan and we need to apply the runtime labels to the assets
+	if runtimeEnv != nil &&
+		runtimeEnv.IsAutomatedEnv() &&
+		inv.Spec.Assets[0].Category == inventory.AssetCategory_CATEGORY_CICD {
+		runtimeLabels = runtimeEnv.Labels()
+	}
+
+	discoveredAssets := &DiscoveredAssets{platformIds: map[string]struct{}{}}
+
+	// we connect and perform discovery for each asset in the job inventory
+	for _, rootAsset := range invAssets {
+		resolvedRootAsset, err := im.ResolveAsset(rootAsset)
+		if err != nil {
+			return nil, err
+		}
+
+		rootRuntime, err := providers.Coordinator.RuntimeFor(resolvedRootAsset, providers.DefaultRuntime())
+		if err != nil {
+			log.Error().Err(err).Str("asset", resolvedRootAsset.Name).Msg("unable to create runtime for asset")
+			discoveredAssets.AddError(resolvedRootAsset, err)
+			continue
+		}
+		if err := rootRuntime.SetRecording(recording); err != nil {
+			discoveredAssets.AddError(resolvedRootAsset, err)
+			continue
+		}
+
+		if err := rootRuntime.Connect(&plugin.ConnectReq{
+			Features: cnquery.GetFeatures(ctx),
+			Asset:    resolvedRootAsset,
+			Upstream: upstream,
+		}); err != nil {
+			log.Error().Err(err).Msg("unable to connect to asset")
+			discoveredAssets.AddError(resolvedRootAsset, err)
+			continue
+		}
+		resolvedRootAsset = rootRuntime.Provider.Connection.Asset // to ensure we get all the information the connect call gave us
+
+		// for all discovered assets, we apply mondoo-specific labels and annotations that come from the root asset
+		for _, a := range rootRuntime.Provider.Connection.Inventory.Spec.Assets {
+			runtime, err := providers.Coordinator.RuntimeFor(resolvedRootAsset, providers.DefaultRuntime())
+			if err != nil {
+				discoveredAssets.AddError(a, err)
+				continue
+			}
+			if err := runtime.SetRecording(recording); err != nil {
+				discoveredAssets.AddError(resolvedRootAsset, err)
+				runtime.Close()
+				continue
+			}
+
+			err = runtime.Connect(&plugin.ConnectReq{
+				Features: config.Features,
+				Asset:    a,
+				Upstream: upstream,
+			})
+			if err != nil {
+				discoveredAssets.AddError(a, err)
+				runtime.Close()
+				continue
+			}
+
+			resolvedAsset := runtime.Provider.Connection.Asset
+			prepareAsset(resolvedAsset, resolvedRootAsset, runtimeLabels)
+
+			// If the asset has been already added, we should close its runtime
+			if !discoveredAssets.Add(resolvedAsset, runtime) {
+				runtime.Close()
```

Comment:
is the idea we keep a runtime open for each asset?

Reply:
yes, that's also how the current code works (apart from the fact that it creates 2 runtimes per asset). The goal is to create the runtime during discovery and close it after the scan has finished